---
title: "Introduction to AlphaSimR"
author: "Chris Gaynor"
date: "`r Sys.Date()`"
output: pdf_document
bibliography: bibliography.bib
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This vignette provides a brief introduction to the AlphaSimR package. The AlphaSimR package was designed to provide all the tools necessary for simulating complex plant and animal breeding programs in R. It is also the successor to the AlphaSim program [@AlphaSim]. 

The basic steps of an AlphaSimR simulation are:

1. Create Founder Haplotypes

2. Set Simulation Parameters

3. Create an Initial Population(s)

4. Simulate a Breeding Program

# Create Founder Haplotypes

The first step in an AlphaSimR simulation is to create a set of founder haplotypes. To do this, run the function `runMacs`. The `runMacs` function uses a built in version of the MaCS program [@MaCS]. The `runMacs` function allows for simulation of haplotypes using several predefined species histories or can be used to create a custom population using MaCS command line arguments. 

# Set Simulation Parameters

Parameters for a simulation are stored in an object of `SimParam-class`. The function `createSimulation` creates this object. The `SimParam-class` is a container for global parameters that will be used by many functions in AlphaSimR. To avoid specifying a `SimParam-class` in all functions that use one, it is recommended that you give your `SimParam-class` the name `SIMPARAM`. All functions that take a `SimParam-class` as an argument will search your global environment for an object called `SIMPARAM` if the argument is `NULL`.

Traits are added to the simulation using one of the following functions: `addTraitA`, `addTraitAD`, `addTraitAG`, or `addTraitADG`. These functions can be called multiple times to run simulations with multiple traits. These functions can also be used to simulate correlated pleiotropic traits. Simulations without any traits are possible as long as you don't use a selection function.

SNP chips are added to the simulation using the `addSnpChip` or `addStructuredSnpChips` functions. As with traits, these functions can be called multiple times to assign multiple SNP chips. SNP chips are not necessary for a simulation, so this step can be skipped.

# Create an Initial Population(s)

The main objects in an AlphaSimR simulation are `Pop-class` objects. These objects represent a population that contains one or more individuals. Many AlphaSimR functions use a `Pop-class` object as an argument and return a `Pop-class` object as a result.

To create your first `Pop-class` object use the function `newPop`. This function should only be called after you have finished all of the above steps.

# Simulate a Breeding Program

Listed below are some functions for carrying out operations on populations.

* Viewing genotypes: `pullSnpGeno`, `pullSnpHaplo`, `pullQtlGeno`, `pullQtlHaplo`, `pullSegSitHaplo`

* Viewing summary data: `meanG`, `meanP`, `varG`, `varP`, `varAD`

* Setting phenotypes: `setPheno`, `calcPheno`, `setPhenoGCA`

* Making selections: `selectInd`, `selectFam`, `selectWithinFam`, `selectMale`, `selectFemale`

* Crossing: `makeCross`, `randCross`, `makeCross2`, `randCross2`,
`self`, `makeDH`

* Genomic selection: `writeRecords`, `setEBV`

# References