---
title: "Introduction to AlphaSimR"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
bibliography: bibliography.bib
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

AlphaSimR is the successor to the AlphaSim software for breeding program simulation [@AlphaSim]. It combines the features of its predecessor with the R software environment to create a flexible and easy-to-use software platform for simulating complex plant and animal breeding programs.

There is no definitive way to construct a simulation in AlphaSimR. This is an intentional design aspect of AlphaSimR, because it frees the user from the constraints of predefined simulation structures. However, most simulations will follow a general structure consisting of these four steps:

1. Create Founder Haplotypes
2. Set Simulation Parameters
3. Model the Breeding Program
4. Examine the Results

The easiest way to learn how AlphaSimR works is to learn about these steps. The easiest way to learn about these steps is to look at an example simulation, so this vignette will introduce AlphaSimR using a small example simulation. This vignette will begin with a description of the breeding program being simulated in this example and continue with a description for how to model each of the above steps for the example breeding program. Included in the descriptions is the code used in the example breeding program simulation. The vignette concludes with the full code for the simulation, including the resulting output.

## Example Breeding Program

The example breeding program presented in this vignette is based on the "historical breeding" portion of the simulation performed by Jenko et al. [-@Jenko15]. The simulation models 20 discrete generations of selection in a simplified animal breeding program. Each generation consists of 1000 animals, of which 500 are male and 500 are female. In each generation the best 25 males are selected on the basis of their genetic value for a single polygenic trait. The selected males are then mated to the females to produce 1000 replacement animals for the next generation.

## Create Founder Haplotypes

The first step in the simulation is to create a set of founder haplotypes. These founder haplotypes will form the genomes of the animals in the first generation of the simulation. Thus, we need to create enough haplotypes to form 1000 animals. We also want to model each animal as having 10 chromosome pairs with 1000 segregating sites per chromosome to serve as QTL controlling the trait under selection. To create the necessary haplotypes, we will use the `runMacs` function. The `runMacs` function is a wrapper for the MaCS software, a coalescent simulation program distributed with AlphaSimR [@MaCS]. The appropriate `runMacs` call for meeting the previously described conditions is given below.

```{r eval=FALSE}
founderPop = runMacs(nInd=1000, nChr=10, segSites=1000)
```

## Set Simulation Parameters

The second step in the simulation is to define the global simulation parameters. We will do this in three small steps. The first of these steps is to initialize an object containing the simulation parameters. The object is initialized with the founder haplotypes created above and the code for doing this is given below.

```{r eval=FALSE}
SP = SimParam$new(founderPop)
```

Note that the output of this function is an object of class `SimParam` and that we have called it `SP`. We recommend always saving your output as `SP`. This is because many AlphaSimR functions take an argument called "simParam" and they have a default value of `NULL` for this argument. If you leave this value as `NULL`, the functions will search your global environment for an object called `SP` and use it as the function's argument. Thus, if you save your `SimParam` output as `SP`, you won't have to specify a value for "simParam" arguments.

Our next step is to define the trait used for selection. We do this below by adding an additive trait controlled by 1000 QTL per chromosome.

```{r eval=FALSE}
SP$addTraitA(nQtlPerChr=1000)
```

The final step is to define how gender is determined. We define gender as systematically determined in the code below. This means populations with an even number of animals will always have equal numbers of males and females.

```{r eval=FALSE}
SP$setGender("yes_sys")
```

## Model the Breeding Program

We are now ready to start modeling the breeding program. When you do this, you need to think about what data you'd like to collect. Here we want to measure the change in both mean and variance over time. We will put this data in a `data.frame` so that it's easy to plot with `ggplot` later.

```{r eval=FALSE}
output = data.frame(cycle=0:20,
                    mean=numeric(21),
                    var=numeric(21))
```

We now need to generate the initial population of animals. This step will take the haplotypes from `founderPop` and the information in `SP` to create an object of `Pop-class`. A `Pop-class` object represents a population of individuals and is one of the most important units in AlphaSimR. This is because many functions will use such objects as an argument. Note that the `Pop-class` can also be used "directly". For example, you can pull individuals out to form new (sub-)populations using `[]` and you can merge populations together using `c()`. You won't be doing any of this here, but these functions are useful if you want to perform a task in AlphaSimR that doesn't have a built in function.
This is one of the features that makes AlphaSimR flexible.

```{r eval=FALSE}
pop = newPop(founderPop)
```

Now we want to start recording the mean and variance in the population by filling in the values for cycle zero.

```{r eval=FALSE}
output$mean[1] = meanG(pop)
output$var[1] = varG(pop)
```

The next step is to model the 20 cycles of selection and mating. AlphaSimR has a host of functions for selection and mating. Here we will use the function `selectCross` that selects a number of males and females and mates them. The easiest way to do this is by writing a loop in R. In this loop, we overwrite `pop` in each cycle, because this is the most efficient use of memory and we have no need for saving each population.

```{r eval=FALSE}
for(cycle in 1:20){
  pop = selectCross(pop=pop, nFemale=500, nMale=25,
                    use="gv", nCrosses=1000)
  
  output$mean[cycle+1] = meanG(pop)
  output$var[cycle+1] = varG(pop)
}
```

## Examine the Results

This section provides code for creating plots for both the mean and variance using the `ggplot2`. This task can also be done with `plot` function in R if you don't want to use `ggplot2`.

```{r eval=FALSE}
ggplot(output,aes(x=cycle,y=mean))+
  geom_line()+
  theme_bw()+
  xlab("Breeding Cycle")+
  ylab("Genetic Gain")

ggplot(output,aes(x=cycle,y=var))+
  geom_line()+
  theme_bw()+
  xlab("Breeding Cycle")+
  ylab("Genetic Variance")+
  expand_limits(y=0)
```


## Full Code

```{r}
library(ggplot2)
library(AlphaSimR)

# Create Founder Haplotypes
founderPop = runMacs(nInd=1000, nChr=10, segSites=1000)

# Set Simulation Parameters
SP = SimParam$new(founderPop)
SP$addTraitA(nQtlPerChr=1000)
SP$setGender("yes_sys")

# Model the Breeding Program
output = data.frame(cycle=0:20,
                    mean=numeric(21),
                    var=numeric(21))

pop = newPop(founderPop)

output$mean[1] = meanG(pop)
output$var[1] = varG(pop)

for(cycle in 1:20){
  pop = selectCross(pop=pop, nFemale=500, nMale=25,
                    use="gv", nCrosses=1000)
  
  output$mean[cycle+1] = meanG(pop)
  output$var[cycle+1] = varG(pop)
}

# Examine the Results
ggplot(output, aes(x=cycle, y=mean))+
  geom_line()+
  theme_bw()+
  xlab("Breeding Cycle")+
  ylab("Genetic Gain")

ggplot(output, aes(x=cycle, y=var))+
  geom_line()+
  theme_bw()+
  xlab("Breeding Cycle")+
  ylab("Genetic Variance")+
  expand_limits(y=0)
```

# References