<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ AlphaSimR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">AlphaSimR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/intro.html">Introduction to AlphaSimR</a>
    </li>
    <li>
      <a href="../articles/traits.html">Traits in AlphaSimR</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/gaynorr/AlphaSimR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="HybridPop-class.html">`[`(<i>&lt;HybridPop&gt;</i>)</a></code> <code><a href="HybridPop-class.html">c(<i>&lt;HybridPop&gt;</i>)</a></code> </p>
        </td>
        <td><p>Hybrid population</p></td>
      </tr><tr><td>
          <p><code><a href="LociMap-class.html">LociMap-class</a></code> </p>
        </td>
        <td><p>Loci metadata</p></td>
      </tr><tr><td>
          <p><code><a href="MapPop-class.html">`[`(<i>&lt;MapPop&gt;</i>)</a></code> <code><a href="MapPop-class.html">c(<i>&lt;MapPop&gt;</i>)</a></code> </p>
        </td>
        <td><p>Raw population with genetic map</p></td>
      </tr><tr><td>
          <p><code><a href="MegaPop-class.html">`[`(<i>&lt;MegaPop&gt;</i>)</a></code> <code><a href="MegaPop-class.html">`[[`(<i>&lt;MegaPop&gt;</i>)</a></code> <code><a href="MegaPop-class.html">c(<i>&lt;MegaPop&gt;</i>)</a></code> </p>
        </td>
        <td><p>Mega-Population</p></td>
      </tr><tr><td>
          <p><code><a href="NamedMapPop-class.html">`[`(<i>&lt;NamedMapPop&gt;</i>)</a></code> <code><a href="NamedMapPop-class.html">c(<i>&lt;NamedMapPop&gt;</i>)</a></code> </p>
        </td>
        <td><p>Raw population with genetic map and id</p></td>
      </tr><tr><td>
          <p><code><a href="Pop-class.html">`[`(<i>&lt;Pop&gt;</i>)</a></code> <code><a href="Pop-class.html">c(<i>&lt;Pop&gt;</i>)</a></code> <code><a href="Pop-class.html">show(<i>&lt;Pop&gt;</i>)</a></code> </p>
        </td>
        <td><p>Population</p></td>
      </tr><tr><td>
          <p><code><a href="RRBLUP.html">RRBLUP()</a></code> </p>
        </td>
        <td><p>RR-BLUP Model</p></td>
      </tr><tr><td>
          <p><code><a href="RRBLUP2.html">RRBLUP2()</a></code> </p>
        </td>
        <td><p>RR-BLUP Model 2</p></td>
      </tr><tr><td>
          <p><code><a href="RRBLUPMemUse.html">RRBLUPMemUse()</a></code> </p>
        </td>
        <td><p>RRBLUP Memory Usage</p></td>
      </tr><tr><td>
          <p><code><a href="RRBLUP_D.html">RRBLUP_D()</a></code> </p>
        </td>
        <td><p>RR-BLUP Model with Dominance</p></td>
      </tr><tr><td>
          <p><code><a href="RRBLUP_D2.html">RRBLUP_D2()</a></code> </p>
        </td>
        <td><p>RR-BLUP with Dominance Model 2</p></td>
      </tr><tr><td>
          <p><code><a href="RRBLUP_GCA.html">RRBLUP_GCA()</a></code> </p>
        </td>
        <td><p>RR-BLUP GCA Model</p></td>
      </tr><tr><td>
          <p><code><a href="RRBLUP_GCA2.html">RRBLUP_GCA2()</a></code> </p>
        </td>
        <td><p>RR-BLUP GCA Model 2</p></td>
      </tr><tr><td>
          <p><code><a href="RRBLUP_SCA.html">RRBLUP_SCA()</a></code> </p>
        </td>
        <td><p>RR-BLUP SCA Model</p></td>
      </tr><tr><td>
          <p><code><a href="RRBLUP_SCA2.html">RRBLUP_SCA2()</a></code> </p>
        </td>
        <td><p>RR-BLUP SCA Model 2</p></td>
      </tr><tr><td>
          <p><code><a href="RRsol-class.html">RRsol-class</a></code> </p>
        </td>
        <td><p>RR-BLUP Solution</p></td>
      </tr><tr><td>
          <p><code><a href="RawPop-class.html">`[`(<i>&lt;RawPop&gt;</i>)</a></code> <code><a href="RawPop-class.html">c(<i>&lt;RawPop&gt;</i>)</a></code> <code><a href="RawPop-class.html">show(<i>&lt;RawPop&gt;</i>)</a></code> </p>
        </td>
        <td><p>Raw Population</p></td>
      </tr><tr><td>
          <p><code><a href="SimParam.html">SimParam</a></code> </p>
        </td>
        <td><p>Simulation parameters</p></td>
      </tr><tr><td>
          <p><code><a href="TraitA-class.html">TraitA-class</a></code> </p>
        </td>
        <td><p>Additive trait</p></td>
      </tr><tr><td>
          <p><code><a href="TraitA2-class.html">TraitA2-class</a></code> </p>
        </td>
        <td><p>Sex specific additive trait</p></td>
      </tr><tr><td>
          <p><code><a href="TraitA2D-class.html">TraitA2D-class</a></code> </p>
        </td>
        <td><p>Sex specific additive and dominance trait</p></td>
      </tr><tr><td>
          <p><code><a href="TraitAD-class.html">TraitAD-class</a></code> </p>
        </td>
        <td><p>Additive and dominance trait</p></td>
      </tr><tr><td>
          <p><code><a href="TraitADE-class.html">TraitADE-class</a></code> </p>
        </td>
        <td><p>Additive, dominance, and epistatic trait</p></td>
      </tr><tr><td>
          <p><code><a href="TraitADEG-class.html">TraitADEG-class</a></code> </p>
        </td>
        <td><p>Additive, dominance, epistasis, and GxE trait</p></td>
      </tr><tr><td>
          <p><code><a href="TraitADG-class.html">TraitADG-class</a></code> </p>
        </td>
        <td><p>Additive, dominance and GxE trait</p></td>
      </tr><tr><td>
          <p><code><a href="TraitAE-class.html">TraitAE-class</a></code> </p>
        </td>
        <td><p>Additive and epistatic trait</p></td>
      </tr><tr><td>
          <p><code><a href="TraitAEG-class.html">TraitAEG-class</a></code> </p>
        </td>
        <td><p>Additive, epistasis and GxE trait</p></td>
      </tr><tr><td>
          <p><code><a href="TraitAG-class.html">TraitAG-class</a></code> </p>
        </td>
        <td><p>Additive and GxE trait</p></td>
      </tr><tr><td>
          <p><code><a href="aa.html">aa()</a></code> </p>
        </td>
        <td><p>Additive-by-additive epistatic deviations</p></td>
      </tr><tr><td>
          <p><code><a href="attrition.html">attrition()</a></code> </p>
        </td>
        <td><p>Lose individuals at random</p></td>
      </tr><tr><td>
          <p><code><a href="bv.html">bv()</a></code> </p>
        </td>
        <td><p>Breeding value</p></td>
      </tr><tr><td>
          <p><code><a href="cChr.html">cChr()</a></code> </p>
        </td>
        <td><p>Combine MapPop chromosomes</p></td>
      </tr><tr><td>
          <p><code><a href="calcGCA.html">calcGCA()</a></code> </p>
        </td>
        <td><p>Calculate GCA</p></td>
      </tr><tr><td>
          <p><code><a href="dd.html">dd()</a></code> </p>
        </td>
        <td><p>Dominance deviations</p></td>
      </tr><tr><td>
          <p><code><a href="dot-newPop.html">.newPop()</a></code> </p>
        </td>
        <td><p>Create new population (internal)</p></td>
      </tr><tr><td>
          <p><code><a href="doubleGenome.html">doubleGenome()</a></code> </p>
        </td>
        <td><p>Double the ploidy of individuals</p></td>
      </tr><tr><td>
          <p><code><a href="ebv.html">ebv()</a></code> </p>
        </td>
        <td><p>Estimated breeding value</p></td>
      </tr><tr><td>
          <p><code><a href="editGenome.html">editGenome()</a></code> </p>
        </td>
        <td><p>Edit genome</p></td>
      </tr><tr><td>
          <p><code><a href="editGenomeTopQtl.html">editGenomeTopQtl()</a></code> </p>
        </td>
        <td><p>Edit genome - the top QTL</p></td>
      </tr><tr><td>
          <p><code><a href="fastRRBLUP.html">fastRRBLUP()</a></code> </p>
        </td>
        <td><p>Fast RR-BLUP</p></td>
      </tr><tr><td>
          <p><code><a href="genParam.html">genParam()</a></code> </p>
        </td>
        <td><p>Sumarize genetic parameters</p></td>
      </tr><tr><td>
          <p><code><a href="genicVarA.html">genicVarA()</a></code> </p>
        </td>
        <td><p>Additive genic variance</p></td>
      </tr><tr><td>
          <p><code><a href="genicVarAA.html">genicVarAA()</a></code> </p>
        </td>
        <td><p>Additive-by-additive genic variance</p></td>
      </tr><tr><td>
          <p><code><a href="genicVarD.html">genicVarD()</a></code> </p>
        </td>
        <td><p>Dominance genic variance</p></td>
      </tr><tr><td>
          <p><code><a href="genicVarG.html">genicVarG()</a></code> </p>
        </td>
        <td><p>Total genic variance</p></td>
      </tr><tr><td>
          <p><code><a href="getMisc.html">getMisc()</a></code> </p>
        </td>
        <td><p>Get miscelaneous information in a population</p></td>
      </tr><tr><td>
          <p><code><a href="getNumThreads.html">getNumThreads()</a></code> </p>
        </td>
        <td><p>Number of available threads</p></td>
      </tr><tr><td>
          <p><code><a href="getQtlMap.html">getQtlMap()</a></code> </p>
        </td>
        <td><p>Get QTL genetic map</p></td>
      </tr><tr><td>
          <p><code><a href="getSnpMap.html">getSnpMap()</a></code> </p>
        </td>
        <td><p>Get SNP genetic map</p></td>
      </tr><tr><td>
          <p><code><a href="gv.html">gv()</a></code> </p>
        </td>
        <td><p>Genetic value</p></td>
      </tr><tr><td>
          <p><code><a href="hybridCross.html">hybridCross()</a></code> </p>
        </td>
        <td><p>Hybrid crossing</p></td>
      </tr><tr><td>
          <p><code><a href="importGenMap.html">importGenMap()</a></code> </p>
        </td>
        <td><p>Import genetic map</p></td>
      </tr><tr><td>
          <p><code><a href="importHaplo.html">importHaplo()</a></code> </p>
        </td>
        <td><p>Import haplotypes</p></td>
      </tr><tr><td>
          <p><code><a href="importInbredGeno.html">importInbredGeno()</a></code> </p>
        </td>
        <td><p>Import inbred, diploid genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="isFemale.html">isFemale()</a></code> <code><a href="isFemale.html">isMale()</a></code> </p>
        </td>
        <td><p>Test if individuals of a population are female or male</p></td>
      </tr><tr><td>
          <p><code><a href="makeCross.html">makeCross()</a></code> </p>
        </td>
        <td><p>Make designed crosses</p></td>
      </tr><tr><td>
          <p><code><a href="makeCross2.html">makeCross2()</a></code> </p>
        </td>
        <td><p>Make designed crosses</p></td>
      </tr><tr><td>
          <p><code><a href="makeDH.html">makeDH()</a></code> </p>
        </td>
        <td><p>Generates DH lines</p></td>
      </tr><tr><td>
          <p><code><a href="meanG.html">meanG()</a></code> </p>
        </td>
        <td><p>Mean genetic values</p></td>
      </tr><tr><td>
          <p><code><a href="meanP.html">meanP()</a></code> </p>
        </td>
        <td><p>Mean phenotypic values</p></td>
      </tr><tr><td>
          <p><code><a href="mergeGenome.html">mergeGenome()</a></code> </p>
        </td>
        <td><p>Combine genomes of individuals</p></td>
      </tr><tr><td>
          <p><code><a href="mergePops.html">mergePops()</a></code> </p>
        </td>
        <td><p>Merge list of populations</p></td>
      </tr><tr><td>
          <p><code><a href="mutate.html">mutate()</a></code> </p>
        </td>
        <td><p>Add Random Mutations</p></td>
      </tr><tr><td>
          <p><code><a href="nInd.html">nInd()</a></code> </p>
        </td>
        <td><p>Number of individuals</p></td>
      </tr><tr><td>
          <p><code><a href="newMapPop.html">newMapPop()</a></code> </p>
        </td>
        <td><p>New MapPop</p></td>
      </tr><tr><td>
          <p><code><a href="newMegaPop.html">newMegaPop()</a></code> </p>
        </td>
        <td><p>Create new Mega Population</p></td>
      </tr><tr><td>
          <p><code><a href="newPop.html">newPop()</a></code> </p>
        </td>
        <td><p>Create new population</p></td>
      </tr><tr><td>
          <p><code><a href="pedigreeCross.html">pedigreeCross()</a></code> </p>
        </td>
        <td><p>Pedigree cross</p></td>
      </tr><tr><td>
          <p><code><a href="pheno.html">pheno()</a></code> </p>
        </td>
        <td><p>Phenotype</p></td>
      </tr><tr><td>
          <p><code><a href="popVar.html">popVar()</a></code> </p>
        </td>
        <td><p>Population variance</p></td>
      </tr><tr><td>
          <p><code><a href="pullIbdHaplo.html">pullIbdHaplo()</a></code> </p>
        </td>
        <td><p>Pull IBD haplotypes</p></td>
      </tr><tr><td>
          <p><code><a href="pullMarkerGeno.html">pullMarkerGeno()</a></code> </p>
        </td>
        <td><p>Pull marker genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="pullMarkerHaplo.html">pullMarkerHaplo()</a></code> </p>
        </td>
        <td><p>Pull marker haplotypes</p></td>
      </tr><tr><td>
          <p><code><a href="pullQtlGeno.html">pullQtlGeno()</a></code> </p>
        </td>
        <td><p>Pull QTL genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="pullQtlHaplo.html">pullQtlHaplo()</a></code> </p>
        </td>
        <td><p>Pull QTL haplotypes</p></td>
      </tr><tr><td>
          <p><code><a href="pullSegSiteGeno.html">pullSegSiteGeno()</a></code> </p>
        </td>
        <td><p>Pull segregating site genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="pullSegSiteHaplo.html">pullSegSiteHaplo()</a></code> </p>
        </td>
        <td><p>Pull seg site haplotypes</p></td>
      </tr><tr><td>
          <p><code><a href="pullSnpGeno.html">pullSnpGeno()</a></code> </p>
        </td>
        <td><p>Pull SNP genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="pullSnpHaplo.html">pullSnpHaplo()</a></code> </p>
        </td>
        <td><p>Pull SNP haplotypes</p></td>
      </tr><tr><td>
          <p><code><a href="quickHaplo.html">quickHaplo()</a></code> </p>
        </td>
        <td><p>Quick founder haplotype simulation</p></td>
      </tr><tr><td>
          <p><code><a href="randCross.html">randCross()</a></code> </p>
        </td>
        <td><p>Make random crosses</p></td>
      </tr><tr><td>
          <p><code><a href="randCross2.html">randCross2()</a></code> </p>
        </td>
        <td><p>Make random crosses</p></td>
      </tr><tr><td>
          <p><code><a href="reduceGenome.html">reduceGenome()</a></code> </p>
        </td>
        <td><p>Create individuals with reduced ploidy</p></td>
      </tr><tr><td>
          <p><code><a href="resetPop.html">resetPop()</a></code> </p>
        </td>
        <td><p>Reset population</p></td>
      </tr><tr><td>
          <p><code><a href="runMacs.html">runMacs()</a></code> </p>
        </td>
        <td><p>Create founder haplotypes using MaCS</p></td>
      </tr><tr><td>
          <p><code><a href="runMacs2.html">runMacs2()</a></code> </p>
        </td>
        <td><p>Alternative wrapper for MaCS</p></td>
      </tr><tr><td>
          <p><code><a href="sampleHaplo.html">sampleHaplo()</a></code> </p>
        </td>
        <td><p>Sample haplotypes from a MapPop</p></td>
      </tr><tr><td>
          <p><code><a href="selIndex.html">selIndex()</a></code> </p>
        </td>
        <td><p>Selection index</p></td>
      </tr><tr><td>
          <p><code><a href="selInt.html">selInt()</a></code> </p>
        </td>
        <td><p>Selection intensity</p></td>
      </tr><tr><td>
          <p><code><a href="selectCross.html">selectCross()</a></code> </p>
        </td>
        <td><p>Select and randomly cross</p></td>
      </tr><tr><td>
          <p><code><a href="selectFam.html">selectFam()</a></code> </p>
        </td>
        <td><p>Select families</p></td>
      </tr><tr><td>
          <p><code><a href="selectInd.html">selectInd()</a></code> </p>
        </td>
        <td><p>Select individuals</p></td>
      </tr><tr><td>
          <p><code><a href="selectOP.html">selectOP()</a></code> </p>
        </td>
        <td><p>Select open pollinating plants</p></td>
      </tr><tr><td>
          <p><code><a href="selectWithinFam.html">selectWithinFam()</a></code> </p>
        </td>
        <td><p>Select individuals within families</p></td>
      </tr><tr><td>
          <p><code><a href="self.html">self()</a></code> </p>
        </td>
        <td><p>Self individuals</p></td>
      </tr><tr><td>
          <p><code><a href="setEBV.html">setEBV()</a></code> </p>
        </td>
        <td><p>Set EBV</p></td>
      </tr><tr><td>
          <p><code><a href="setMisc.html">setMisc()</a></code> </p>
        </td>
        <td><p>Set miscelaneous information in a population</p></td>
      </tr><tr><td>
          <p><code><a href="setPheno.html">setPheno()</a></code> </p>
        </td>
        <td><p>Set phenotypes</p></td>
      </tr><tr><td>
          <p><code><a href="setPhenoGCA.html">setPhenoGCA()</a></code> </p>
        </td>
        <td><p>Set GCA as phenotype</p></td>
      </tr><tr><td>
          <p><code><a href="setPhenoProgTest.html">setPhenoProgTest()</a></code> </p>
        </td>
        <td><p>Set progeny test as phenotype</p></td>
      </tr><tr><td>
          <p><code><a href="smithHazel.html">smithHazel()</a></code> </p>
        </td>
        <td><p>Calculate Smith-Hazel weights</p></td>
      </tr><tr><td>
          <p><code><a href="solveMKM.html">solveMKM()</a></code> </p>
        </td>
        <td><p>Solve Multikernel Model</p></td>
      </tr><tr><td>
          <p><code><a href="solveMVM.html">solveMVM()</a></code> </p>
        </td>
        <td><p>Solve Multivariate Model</p></td>
      </tr><tr><td>
          <p><code><a href="solveRRBLUP.html">solveRRBLUP()</a></code> </p>
        </td>
        <td><p>Solve RR-BLUP</p></td>
      </tr><tr><td>
          <p><code><a href="solveRRBLUPMK.html">solveRRBLUPMK()</a></code> </p>
        </td>
        <td><p>Solve Multikernel RR-BLUP</p></td>
      </tr><tr><td>
          <p><code><a href="solveRRBLUPMV.html">solveRRBLUPMV()</a></code> </p>
        </td>
        <td><p>Solve Multivariate RR-BLUP</p></td>
      </tr><tr><td>
          <p><code><a href="solveRRBLUP_EM.html">solveRRBLUP_EM()</a></code> </p>
        </td>
        <td><p>Solve RR-BLUP with EM</p></td>
      </tr><tr><td>
          <p><code><a href="solveRRBLUP_EM2.html">solveRRBLUP_EM2()</a></code> </p>
        </td>
        <td><p>Solve RR-BLUP with EM and 2 random effects</p></td>
      </tr><tr><td>
          <p><code><a href="solveRRBLUP_EM3.html">solveRRBLUP_EM3()</a></code> </p>
        </td>
        <td><p>Solve RR-BLUP with EM and 3 random effects</p></td>
      </tr><tr><td>
          <p><code><a href="solveUVM.html">solveUVM()</a></code> </p>
        </td>
        <td><p>Solve Univariate Model</p></td>
      </tr><tr><td>
          <p><code><a href="transMat.html">transMat()</a></code> </p>
        </td>
        <td><p>Linear transformation matrix</p></td>
      </tr><tr><td>
          <p><code><a href="usefulness.html">usefulness()</a></code> </p>
        </td>
        <td><p>Usefulness criterion</p></td>
      </tr><tr><td>
          <p><code><a href="varA.html">varA()</a></code> </p>
        </td>
        <td><p>Additive variance</p></td>
      </tr><tr><td>
          <p><code><a href="varAA.html">varAA()</a></code> </p>
        </td>
        <td><p>Additive-by-additive epistatic variance</p></td>
      </tr><tr><td>
          <p><code><a href="varD.html">varD()</a></code> </p>
        </td>
        <td><p>Dominance variance</p></td>
      </tr><tr><td>
          <p><code><a href="varG.html">varG()</a></code> </p>
        </td>
        <td><p>Total genetic variance</p></td>
      </tr><tr><td>
          <p><code><a href="varP.html">varP()</a></code> </p>
        </td>
        <td><p>Phenotypic variance</p></td>
      </tr><tr><td>
          <p><code><a href="writePlink.html">writePlink()</a></code> </p>
        </td>
        <td><p>Writes a Pop-class as PLINK files</p></td>
      </tr><tr><td>
          <p><code><a href="writeRecords.html">writeRecords()</a></code> </p>
        </td>
        <td><p>Write data records</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Chris Gaynor.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

      </footer></div>

  


  

  </body></html>

