<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Chris Gaynor" />

<meta name="date" content="2017-11-22" />

<title>Introduction to AlphaSimR</title>






<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>

</head>

<body>




<h1 class="title toc-ignore">Introduction to AlphaSimR</h1>
<h4 class="author"><em>Chris Gaynor</em></h4>
<h4 class="date"><em>2017-11-22</em></h4>



<p>The AlphaSimR package provides a framework for simulating plant and animal breeding programs in R. It is based on the <a href="https://dl.sciencesocieties.org/publications/tpg/articles/0/0/plantgenome2016.02.0013">AlphaSim</a> program. This document will introduce the basic steps for running a simulation in AlphaSimR.</p>
<p>The basic steps of an AlphaSimR simulation are:</p>
<ol type="1">
<li><p>Create founder haplotypes using <code>runMacs</code></p></li>
<li><p>Set simulation parameters using <code>createSimulation</code></p></li>
<li><p>Add traits using <code>addTraitA</code> or any other addTrait function</p></li>
<li><p>Optionally, add SNP chips using <code>addSnpChip</code></p></li>
<li><p>Create one or more initial populations using <code>newPop</code></p></li>
<li><p>Perform a mix of crossing and selection to simulate a breeding program</p></li>
</ol>
<section id="creating-founder-haplotypes" class="level3">
<h3>Creating founder haplotypes</h3>
<p>The first step to a simulation in AlphaSimR is to create a set of founder haplotypes. This is done using the <code>runMacs</code> function which uses a built in version of the <a href="https://github.com/gchen98/macs">MaCS</a> program. The <code>runMacs</code> function allows for simulation of haplotypes using several predefined species histories or can be used to run a custom population using MaCS command line arguements.</p>
</section>
<section id="setting-simulation-parameters" class="level3">
<h3>Setting simulation parameters</h3>
<p>Parameters for a simulation are stored in an object of <code>SimParam-class</code>. This object is created by calling the function <code>createSimulation</code>. The <code>SimParam-class</code> is a container for global parameters that will be used by many functions in AlphaSimR. To avoid specifying a <code>SimParam-class</code> in all functions that use one, it is recommended that you give your <code>SimParam-class</code> the name <code>SIMPARAM</code>. All functions that take a <code>SimParam-class</code> as an arguement will search your global environment for an object called <code>SIMPARAM</code> if the argument is <code>NULL</code>.</p>
</section>
<section id="adding-traits" class="level3">
<h3>Adding traits</h3>
<p>Traits are added to the simulation using one of the fullowing functions: <code>addTraitA</code>, <code>addTraitAD</code>, <code>addTraitAG</code>, or <code>addTraitADG</code>. These functions can be called multiple times to run simulations with multiple traits. These functions can also be used to simulate correlated pleiotropic traits. Simulation without any traits are possible as long as you donâ€™t use a selection function.</p>
</section>
<section id="adding-snp-chips" class="level3">
<h3>Adding SNP chips</h3>
<p>SNP chips are added to the simulation using the <code>addSnpChip</code> or <code>addStructuredSnpChips</code> functions. As with traits, these functions can be called multiple times to assign multiple SNP chips. SNP chips are not necessary for a simulation, so this step can be skipped.</p>
</section>
<section id="creating-a-population" class="level3">
<h3>Creating a population</h3>
<p>The main objects in an AlphaSimR simulation are <code>Pop-class</code> objects. These objects represent a population which contains one or more individuals. Many AlphaSimR functions use a <code>Pop-class</code> object as an argument and return a <code>Pop-class</code> object as a result.</p>
<p>To create your first <code>Pop-class</code> object use the function <code>newPop</code>. This function should only be called after you have finished all of the above steps.</p>
</section>
<section id="working-with-populations" class="level3">
<h3>Working with populations</h3>
<p>Below are some functions for operations carried out on populations.</p>
<ul>
<li><p>Viewing genotypes: <code>pullSnpGeno</code>, <code>pullSnpHaplo</code>, <code>pullQtlGeno</code>, <code>pullQtlHaplo</code>, <code>pullSegSitHaplo</code></p></li>
<li><p>Viewing summary data: <code>meanG</code>, <code>meanP</code>, <code>varG</code>, <code>varP</code>, <code>varAD</code></p></li>
<li><p>Setting phenotypes: <code>setPheno</code>, <code>calcPheno</code>, <code>setPhenoGCA</code></p></li>
<li><p>Making selections: <code>selectInd</code>, <code>selectFam</code>, <code>selectWithinFam</code>, <code>selectMale</code>, <code>selectFemale</code></p></li>
<li><p>Crossing: <code>makeCross</code>, <code>randCross</code>, <code>makeCross2</code>, <code>randCross2</code>, <code>self</code>, <code>makeDH</code></p></li>
<li><p>Genomic selection: <code>writeRecords</code>, <code>setEBV</code></p></li>
</ul>
</section>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
