<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Pull QTL haplotypes — pullQtlHaplo • AlphaSimR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Pull QTL haplotypes — pullQtlHaplo" />
<meta property="og:description" content="Retrieves QTL haplotype data" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">AlphaSimR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/intro.html">Introduction to AlphaSimR</a>
    </li>
    <li>
      <a href="../articles/traits.html">Traits in AlphaSimR</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/gaynorr/AlphaSimR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Pull QTL haplotypes</h1>
    <small class="dont-index">Source: <a href='https://github.com/gaynorr/AlphaSimR/blob/master/R/pullGeno.R'><code>R/pullGeno.R</code></a></small>
    <div class="hidden name"><code>pullQtlHaplo.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Retrieves QTL haplotype data</p>
    </div>

    <pre class="usage"><span class='fu'>pullQtlHaplo</span><span class='op'>(</span><span class='va'>pop</span>, trait <span class='op'>=</span> <span class='fl'>1</span>, haplo <span class='op'>=</span> <span class='st'>"all"</span>, chr <span class='op'>=</span> <span class='cn'>NULL</span>, simParam <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pop</th>
      <td><p>an object of <code><a href='Pop-class.html'>Pop-class</a></code></p></td>
    </tr>
    <tr>
      <th>trait</th>
      <td><p>an integer. Indicates which trait's
QTL haplotypes to retrieve.</p></td>
    </tr>
    <tr>
      <th>haplo</th>
      <td><p>either "all" for all haplotypes or an integer
for a single set of haplotypes. Use a value of 1 for female
haplotyes and a value of 2 for male haplotypes.</p></td>
    </tr>
    <tr>
      <th>chr</th>
      <td><p>a vector of chromosomes to retrieve. If NULL,
all chromosome are retrieved.</p></td>
    </tr>
    <tr>
      <th>simParam</th>
      <td><p>an object of <code><a href='SimParam.html'>SimParam</a></code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns a matrix of QTL haplotypes.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>#' @title Pull SNP haplotypes for multiple chips
#'
#' @description Retrieves SNP haplotype data for multiple snp
#'
#' @param pop an object of <code><a href='Pop-class.html'>Pop-class</a></code>
#' @param chips a vector. For each animal indicates what snp
#' chip to use
#' @param haplo either "all" for all haplotypes or an integer
#' for a single set of haplotypes. Use a value of 1 for female
#' haplotyes and a value of 2 for male haplotypes.
#' @param missing What value to use for missing
#' @param simParam an object of <code><a href='SimParam.html'>SimParam</a></code>
#'
#' @return Returns a matrix of SNP haplotypes.
#' 
#' @export
pullMultipleSnpHaplo = function(pop, chips, haplo="all",
                                missing=9, simParam=NULL)
  if(is.null(simParam))
    simParam = get("SP",envir=.GlobalEnv)  # I feel like the next line shouldn't be needed but I don't know
  # enough R! (dmoney)
  missing = as.integer(missing)
  allSnps = numeric(0)
  uniqueChips = unique(chips)
  for (c in uniqueChips)
    allSnps = sort(union(allSnps,simParam$snpChips[[c]]@lociLoc))  if (haplo == "all") 
    output = matrix(pop@nInd*2,length(allSnps),data=missing)
    if(class(pop)=="Pop")
      rownames(output) = paste(rep(pop@id,each=pop@ploidy),
                               rep(1:pop@ploidy,pop@nInd),sep="_")
    else
      rownames(output) = paste(rep(1:pop@nInd,each=pop@ploidy),
                               rep(1:pop@ploidy,pop@nInd),sep="_")  else
    output = matrix(pop@nInd,length(allSnps),data=missing)
    if(class(pop)=="Pop")
      rownames(output) = paste(pop@id,rep(haplo,pop@nInd),sep="_")
    else
      rownames(output) = paste(1:pop@nInd,rep(haplo,pop@nInd),sep="_")  for (snpChip in uniqueChips)
    mask = allSnps <!-- %in% simParam$snpChips[[snpChip]]@lociLoc -->
    if (haplo == "all") 
      one = getHaplo(pop@geno,
                     simParam$snpChips[[snpChip]]@lociPerChr,
                     simParam$snpChips[[snpChip]]@lociLoc,
                     simParam$nThreads)
      one = convToImat(one)
      for (i in 1:pop@nInd)
        if (chips[i] == snpChip) 
          output[i*2-1,mask] = one[i*2-1,]
          output[i*2,mask] = one[i*2,]    else 
      one = getOneHaplo(pop@geno,
                        simParam$snpChips[[snpChip]]@lociPerChr,
                        simParam$snpChips[[snpChip]]@lociLoc,
                        as.integer(haplo),
                        simParam$nThreads)
      one = convToImat(one)
      for (i in 1:pop@nInd)
        if (chips[i] == snpChip) 
          output[i,mask] = one[i,]
          output[i,mask] = one[i,]  colnames(output) = paste("SNP",1:ncol(output),sep="_")  return(output)</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>#Create founder haplotypes</span>
<span class='va'>founderPop</span> <span class='op'>=</span> <span class='fu'><a href='quickHaplo.html'>quickHaplo</a></span><span class='op'>(</span>nInd<span class='op'>=</span><span class='fl'>10</span>, nChr<span class='op'>=</span><span class='fl'>1</span>, segSites<span class='op'>=</span><span class='fl'>15</span><span class='op'>)</span>

<span class='co'>#Set simulation parameters</span>
<span class='va'>SP</span> <span class='op'>=</span> <span class='va'><a href='SimParam.html'>SimParam</a></span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span><span class='va'>founderPop</span><span class='op'>)</span>
<span class='va'>SP</span><span class='op'>$</span><span class='fu'>addTraitA</span><span class='op'>(</span><span class='fl'>10</span><span class='op'>)</span>
<span class='va'>SP</span><span class='op'>$</span><span class='fu'>addSnpChip</span><span class='op'>(</span><span class='fl'>5</span><span class='op'>)</span>

<span class='co'>#Create population</span>
<span class='va'>pop</span> <span class='op'>=</span> <span class='fu'><a href='newPop.html'>newPop</a></span><span class='op'>(</span><span class='va'>founderPop</span>, simParam<span class='op'>=</span><span class='va'>SP</span><span class='op'>)</span>
<span class='fu'>pullQtlHaplo</span><span class='op'>(</span><span class='va'>pop</span>, simParam<span class='op'>=</span><span class='va'>SP</span><span class='op'>)</span>
</div><div class='output co'>#&gt;      QTL_1 QTL_2 QTL_3 QTL_4 QTL_5 QTL_6 QTL_7 QTL_8 QTL_9 QTL_10
#&gt; 1_1      0     1     0     0     1     1     0     0     0      1
#&gt; 1_2      0     0     1     1     1     0     1     1     0      0
#&gt; 2_1      1     0     0     1     1     0     1     0     1      0
#&gt; 2_2      0     0     0     0     1     0     0     0     0      0
#&gt; 3_1      0     1     1     0     0     1     1     1     1      0
#&gt; 3_2      1     1     1     1     0     0     1     1     0      1
#&gt; 4_1      1     0     0     0     1     1     0     1     1      0
#&gt; 4_2      0     0     1     0     0     0     1     0     1      1
#&gt; 5_1      1     1     1     1     1     0     0     0     1      1
#&gt; 5_2      1     1     1     0     1     1     0     1     1      0
#&gt; 6_1      0     0     1     1     0     0     0     0     1      1
#&gt; 6_2      1     1     1     0     1     0     1     0     1      0
#&gt; 7_1      0     0     1     0     1     1     0     1     1      1
#&gt; 7_2      1     1     0     1     0     1     1     1     1      0
#&gt; 8_1      1     1     1     0     1     1     1     1     0      1
#&gt; 8_2      0     1     0     0     0     1     0     0     0      1
#&gt; 9_1      1     1     1     0     0     0     1     1     1      1
#&gt; 9_2      0     1     1     0     1     1     1     0     0      1
#&gt; 10_1     0     1     0     0     1     1     1     1     0      1
#&gt; 10_2     0     0     1     0     0     0     1     0     0      0</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Chris Gaynor.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


