#Class RRBLUPsolution----
#' @title RR-BLUP Solution
#' 
#' @description 
#' Contains estimated marker effects generated by \code{\link{RRBLUP}}.
#' 
#' @slot markerEffects list of estimated marker effects, stored as 
#' \code{\link{TraitA-class}}
#' @slot intercept vector of intercepts from RR-BLUP solutions
#'
#' @export
setClass("RRBLUPsolution",
         slots=c(markerEffects="list",
                 intercept="numeric"))

setValidity("RRBLUPsolution",function(object){
  errors = character()
  if(length(object@markerEffects)!=length(object@intercept)){
    errors = c(errors,"length(markerEffects)!=length(intercept)")
  }
  if(length(errors)==0){
    return(TRUE)
  }else{
    return(errors)
  }
})

#' @title Predict GEBV
#' 
#' @description 
#' Predicts GEBVs for a population using marker effects 
#' estimated by \code{\link{RRBLUP}}.
#' 
#' @param object an object of \code{\link{RRBLUPsolution-class}}
#' @param pop an object of \code{\link{Pop-class}}
#' 
#' @return an object of \code{\link{Pop-class}} containing the 
#' estimated GEBVs.
#' 
#' @export
#' @docType methods
setMethod("predict",
          signature(object = "RRBLUPsolution"),
          function(object, pop, ...){
            ebv = lapply(object@markerEffects,getGv,pop=pop,w=0)
            ebv = do.call("cbind",ebv)
            pop@ebv = ebv
            return(pop)
          }
)

# GS functions----

#' @title Write data records
#' 
#' @description
#' Saves a population's phenotypic and marker data to a directory.
#'
#' @param pop an object of \code{\link{Pop-class}}
#' @param dir path to a directory for saving output
#' @param snpChip which SNP chip genotype to save. If useQtl=TRUE, this 
#' value will indicate which trait's QTL genotype to save. A value of 
#' 0 will skip writing a snpChip.
#' @param useQtl should QTL genotype be written instead of SNP chip 
#' genotypes.
#' @param reps number of reps for phenotypes. This values is used for modelling 
#' heterogenous error variance in genomic selection models. Leave value as 1 
#' unless using reps for phenotypes.
#' @param fixEff an integer indicating levels of fixed effect. Leave 
#' value as 1 if not using different levels of fixed effects.
#' @param simParam an object of \code{\link{SimParam-class}}
#'
#' @export
writeRecords = function(pop,dir,snpChip,useQtl=FALSE,reps=1,fixEff=1,
                        simParam=SIMPARAM){
  stopifnot(dir.exists(dir))
  if(snpChip==0){
    nMarkers = 0
    markerType = "NULL"
  }else{
    if(useQtl){
      nMarkers = simParam@traits[[snpChip]]@nLoci
      markerType = paste0("QTL_",snpChip)
    }else{
      nMarkers = simParam@snpChips[[snpChip]]@nLoci
      markerType = paste0("SNP_",snpChip)
    }
  }
  #Check that the marker set isn't being changed
  nMarkerPath = file.path(dir,"nMarkers.txt")
  if(file.exists(nMarkerPath)){
    nMarkersDir = scan(nMarkerPath,integer(),quiet=TRUE)
    stopifnot(nMarkersDir==nMarkers)
  }else{
    writeLines(as.character(nMarkers),nMarkerPath)
  }
  markerTypePath = file.path(dir,"markerType.txt")
  if(file.exists(markerTypePath)){
    markerTypeDir = scan(markerTypePath,character(),quiet=TRUE)
    stopifnot(markerTypeDir==markerType)
  }else{
    writeLines(markerType,markerTypePath)
  }
  #Write info.txt
  info = data.frame(id=pop@id,mother=pop@mother,father=pop@father,
                    reps=rep(reps,pop@nInd),fixEff=rep(fixEff,pop@nInd),
                    stringsAsFactors=FALSE)
  filePath = file.path(dir,"info.txt")
  if(file.exists(filePath)){
    write.table(info,filePath,append=TRUE,col.names=FALSE,
                row.names=FALSE)
  }else{
    write.table(info,filePath,row.names=FALSE)
  }
  #Write gv.txt
  write.table(pop@gv,file.path(dir,"gv.txt"),append=TRUE,
              col.names=FALSE,row.names=FALSE)
  #Write pheno.txt
  write.table(pop@pheno,file.path(dir,"pheno.txt"),append=TRUE,
              col.names=FALSE,row.names=FALSE)
  #Write markers.txt, unless snpChip=0
  if(snpChip!=0){
    if(useQtl){
      write.table(pullQtlGeno(pop,snpChip,simParam=simParam),
                  file.path(dir,"markers.txt"),append=TRUE,
                  col.names=FALSE,row.names=FALSE)
    }else{
      write.table(pullSnpGeno(pop,snpChip,simParam=simParam),
                  file.path(dir,"markers.txt"),append=TRUE,
                  col.names=FALSE,row.names=FALSE)
    }
  }
}

#' @title RR-BLUP Model
#' 
#' @description
#' Fits an RR-BLUP model for subsequent prediction.
#'
#' @param dir path to a directory with output from /code{/link{writeRecords}}
#' @param traits an integer indicating the trait or traits to model, or a 
#' function of the traits returning a single value.
#' @param useGv should genetic values be used instead of phenotypes
#' @param simParam an object of \code{\link{SimParam-class}}
#'
#' @export
RRBLUP = function(dir,traits=1,useGv=FALSE, simParam=SIMPARAM){
  #Read and calculate basic information
  markerInfo = read.table(file.path(dir,"info.txt"),header=TRUE,
                          comment.char="",stringsAsFactors=FALSE)
  nInd = nrow(markerInfo)
  nMarkers = scan(file.path(dir,"nMarkers.txt"),integer(),quiet=TRUE)
  markerType =scan(file.path(dir,"markerType.txt"),character(),quiet=TRUE)
  #Set trait/traits for genomic selection
  if(useGv){
    y = scan(file.path(dir,"gv.txt"),numeric(),quiet=TRUE)
  }else{
    y = scan(file.path(dir,"pheno.txt"),numeric(),quiet=TRUE)
  }
  y = matrix(y,nrow=nInd,ncol=length(y)/nInd,byrow=TRUE)
  if(is.function(traits)){
    y = apply(y,1,traits)
    y = as.matrix(y)
  }else{
    y = y[,traits,drop=FALSE]
  }
  #Fit model
  if(ncol(y)>1){
    output = callRRBLUP_MV(y,markerInfo$fixEff,markerInfo$reps,
                           file.path(dir,"markers.txt"),nMarkers)
  }else{
    output = callRRBLUP(y,markerInfo$fixEff,markerInfo$reps,
                        file.path(dir,"markers.txt"),nMarkers)
  }
  #ToDo
  tmp = unlist(strsplit(markerType,"_"))
  if(tmp[1]=="SNP"){
    markerInfo = simParam@snpChips[[as.numeric(tmp[2])]]
  }else{
    markerInfo = simParam@traits[[as.numeric(tmp[2])]]
  }
  markerEffects = list()
  for(i in 1:ncol(output$u)){
    markerEffects[[i]] = new("TraitA",
                             nLoci=markerInfo@nLoci,
                             lociPerChr=markerInfo@lociPerChr,
                             lociLoc=markerInfo@lociLoc,
                             addEff=output$u[,i],
                             intercept=0.0)
  }
  intercept = output$beta[1,]
  output = new("RRBLUPsolution",
               markerEffects=markerEffects,
               intercept=intercept)
  return(output)
}

#' @title SolveMKM Objective
#' 
#' @description
#' Used internally by SolveMKM and not for to be used directly.
#' 
#' @param x a numeric vector
#' @param ptrData a pointer
#' 
#' @export
objWeightsR = function(x, ptrData){
  tmp = objWeights(x,ptrData)
  return(tmp)
}

#' @title Single Cross GBLUP
#' 
#' @description
#' Saves a population's phenotypic and marker data to a directory.
#'
#' @param phenoDir path to a directory with hybrid phenotypes or genetic values. 
#' The directory must contain output from /code{/link{writeRecords}} 
#' @param femaleDir path to a directory with female genotype data. 
#' The directory must contain output from /code{/link{writeRecords}} 
#' @param maleDir path to a directory with male genotype data. 
#' The directory must contain output from /code{/link{writeRecords}} 
#' @param trait an integer indicating the trait to model, or a 
#' function of the traits returning a single value.
#' @param useGv should genetic values be used instead of phenotypes
#'
#' @export
SC_GBLUP = function(phenoDir,femaleDir,maleDir,trait=1,useGv=FALSE){
  #Read and calculate basic information
  markerInfo = read.table(file.path(phenoDir,"info.txt"),header=TRUE,
                          comment.char="",stringsAsFactors=FALSE)
  femaleInfo = read.table(file.path(phenoDir,"info.txt"),header=TRUE,
                          comment.char="",stringsAsFactors=FALSE)
  maleInfo = read.table(file.path(phenoDir,"info.txt"),header=TRUE,
                        comment.char="",stringsAsFactors=FALSE)
  nInd = nrow(markerInfo)
  nFemale = nrow(femaleInfo)
  nMale = nrow(maleInfo)
  nMarkers = scan(file.path(femaleDir,"nMarkers.txt"),integer(),quiet=TRUE)
  tmp = scan(file.path(maleDir,"nMarkers.txt"),integer(),quiet=TRUE)
  stopifnot(nMarkers==tmp)
  markerType = scan(file.path(femaleDir,"markerType.txt"),character(),quiet=TRUE)
  tmp = scan(file.path(maleDir,"markerType.txt"),character(),quiet=TRUE)
  stopifnot(markerType==tmp)
  #Set trait/traits for genomic selection
  if(useGv){
    y = scan(file.path(dir,"gv.txt"),numeric(),quiet=TRUE)
  }else{
    y = scan(file.path(dir,"pheno.txt"),numeric(),quiet=TRUE)
  }
  y = matrix(y,nrow=nInd,ncol=length(y)/nInd,byrow=TRUE)
  if(is.function(traits)){
    y = apply(y,1,traits)
    y = as.matrix(y)
  }else{
    y = y[,trait,drop=FALSE]
  }
  #Fit model
  ## ToDo
  #output = callRRBLUP(y,markerInfo$fixEff,markerInfo$reps,
  #                    file.path(dir,"markers.txt"),nMarkers)
  output["markerType"] = markerType
  class(output) = "SC_GBLUP Solution"
  return(output)
}